src/FSharpPlus/DList.fs

```diff
@@ -1,45 +1,260 @@
+    member this.TryHead = DList<'T>.tryHead data
+
+    ///O(1). Returns true if the DList has no elements.
+    member this.IsEmpty = match data with Nil -> true | _ -> false
+
+    ///O(1). Returns a new DList with the element added to the end.
+    member this.Conj (x:'T) = DList( (length + 1), DList<'T>.append(data, Unit x) )
+
+    ///O(log n). Returns a new DList of the elements trailing the first element.
+    member this.Tail =
+        let rec step (xs:DListData<'T>) (acc:DListData<'T>) =
+            match xs with
+            | Nil -> acc
+            | Unit _ -> acc
+            | Join(x,y) -> step x (DList<'T>.append(y, acc))
+        if this.IsEmpty then failwith "DList.tail: empty DList"
+        else DList( (length - 1), (step data Nil ))
+
+    ///O(log n). Returns option DList of the elements trailing the first element.
+    member this.TryTail =
+        let rec step (xs:DListData<'T>) (acc:DListData<'T>) =
+            match xs with
+            | Nil -> acc
+            | Unit _ -> acc
+            | Join(x,y) -> step x (DList<'T>.append(y, acc))
+        if this.IsEmpty then None
+        else Some (DList( (length - 1), (step data Nil )))
+
+    ///O(log n). Returns the first element and tail.
+    member this.Uncons = ((DList<'T>.head data), (this.Tail))
+
+    ///O(log n). Returns option first element and tail.
+    member this.TryUncons =
+        match DList<'T>.tryHead data with
+        | Some(x) -> Some (x, this.Tail)
+        | None -> None
+
+    member this.toSeq() =
+        //adaptation of right-hand side of Norman Ramsey's "fold"
+        let rec walk rights l = 
+           seq {match l with
+                | Nil       -> 
+                    match rights with
+                    | []    -> () 
+                    | t::ts -> yield! walk ts t 
+                | Unit x    -> 
+                    yield x
+                    match rights with
+                    | []    -> () 
+                    | t::ts -> yield! walk ts t 
+                | Join(x,y) -> yield! walk (y::rights) x}
+               
+        (walk [] data).GetEnumerator()
+
+    interface IEnumerable<'T> with
+        member s.GetEnumerator() = s.toSeq()
+
+    interface System.Collections.IEnumerable with
+        override s.GetEnumerator() = (s.toSeq() :> System.Collections.IEnumerator)
+            
+and 
+    DListData<'T> =
+    | Nil
+    | Unit of 'T
+    | Join of DListData<'T> * DListData<'T>  
+
+[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module DList =
-    let empty = DList id
-    let toList (DList x)   = x []
-    let toSeq  (DList x)   = x [] :> seq<_>
-    let ofList source = source |> List.append |> DList
-    let ofSeq  source = source |> Seq.toList |> List.append |> DList
-    let append (DList x) (DList y) = DList (x << y)
-    let singleton x = DList (fun r -> x::r)
-    let cons x (DList f) = (DList (fun l -> x::(f l)))
-    let snoc (DList f) x = (DList (fun l -> f (x::l)))
```

Comment:
`snoc` has changed name to `conj`, this is a breaking change for any user of F#+

Reply:
Yes, it is. Anyway I was going to suggest to use ``add``.

Reply:
I'll rename it to add.